pkg/apis/core/validation/shoot.go

```diff
@@ -1046,12 +1049,31 @@ func ValidateWorkers(workers []core.Worker, fldPath *field.Path) field.ErrorList
 			allErrs = append(allErrs, field.Duplicate(fldPath.Index(i).Child("name"), worker.Name))
 		}
 		workerNames[worker.Name] = true
+
+		if len(worker.Taints) == 0 {
+			atLeastOnePoolWithCompatibleTaints = true
+		} else {
+			onlyPreferNoScheduleEffectTaints := true
+			for _, taint := range worker.Taints {
+				if taint.Effect != corev1.TaintEffectPreferNoSchedule {
+					onlyPreferNoScheduleEffectTaints = false
+					break
+				}
+			}
+			if onlyPreferNoScheduleEffectTaints {
+				atLeastOnePoolWithCompatibleTaints = true
+			}
+		}
```

Comment:
```suggestion
		} else if len(worker.Taints) == 1 && worker.Tains[0].Effect ==  corev1.TaintEffectPreferNoSchedule {
		        atLeastOnePoolWithCompatibleTaints = true
		}
```

Reply:
Why do we need to handle this case specially? It's covered by the general case already.

Reply:
It replaces the case entirely to get rid of the for loop.

Reply:
But now it's resolved anyway since it's been moved to a switch case.

Reply:
I don't get it, sorry. What if I have `NoExecute` taints in the first three pools, the fourth one has `PreferNoSchedule` and the next five pools have `NoExecute` as well?

Reply:
Agree. Sorry for the confusion.

Reply:
As discussed, will add more test cases.